### PR TITLE
Lighter InnerBlocks: allow blocks to render own wrappers

### LINKED ISSUE #7770: Columns block styles needs polishing and probably markup changes in the editor
**Describe the bug**

Columns block styles needs polishing in the front-end and in the editor. Responsive styles for starters.

But a bigger issue that I noticed is the markup.

**Front-end markup**

```html
<div class="wp-block-columns has-2-columns">
   <div class="wp-block-column">
      <p>Cras justo odio, dapibus ac facilisis in, egestas eget quam. Sed posuere consectetur est at lobortis. Nullam id dolor id nibh ultricies vehicula ut id elit. Donec sed odio dui. Nulla vitae elit libero, a pharetra augue.</p>
   </div>
	
   <div class="wp-block-column">
      <p>Cras justo odio, dapibus ac facilisis in, egestas eget quam. Sed posuere consectetur est at lobortis. Nullam id dolor id nibh ultricies vehicula ut id elit. Donec sed odio dui. Nulla vitae elit libero, a pharetra augue.</p>
   </div>
</div>
```

This is now OK for basic columns. And we can style columns as we want.

**Example CSS**
```css
.wp-core-columns {
   display: grid;
   grid-gap: 1rem;
   grid-template-columns: repeat(2, 1fr);

   // etc.
}
```

In the editor the markup looks like this:

```html
<div class="wp-block-columns has-2-columns">
   <div class="editor-inner-blocks">
      <div class="editor-block-list__layout">
         <div class="editor-block-list__block">
            <p>Blocks somewhere over here.</p>
         </div>

         <div class="editor-block-list__block">
            <p>Blocks somewhere over here.</p>
         </div>
      </div>
   </div>
</div>
```
Which means that `wp-block-columns` always have only one direct wrapper `<div class="editor-inner-blocks">` inside it and there is no `<div class="wp-block-column">` anywhere.

**Expected behavior**
Have `<div class="wp-block-column">` wrappers in the editor also so that styles would work.

**Screenshots**
With example CSS columns in the editor looks like this:

columns block: has 2 columns

It's in 2 columns because of Core styles but other than that it doesn't respect any of my styles.

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

closes #7770

Removes 6 wrapper `div` elements from the columns and column together.

I created this PR before the merge of #19701 to illustrate with the columns block how it all comes together. This PR really shows the power that blocks will have when they can entirely render themselves.

With this PR, the DOM tree of the columns block is exactly the same as the DOM tree on the front end. This removes the need add any styling for the editor. There's still a few things that ideally should be adjusted (we shouldn't be adding a margin to blocks by default because some blocks need to reset this). But you get the idea and the direction.

I believe this direction will be really important for full site editing themes and performance as well (which is built on `InnerBlocks`).

It will also allow us to build complex block with inner blocks that need a semantic relationship with the ancestors. Think about a list, table, or definition list built entirely with `InnerBlocks`.

Please note that I haven't polished the styling of the columns block yet.

Before:

Screenshot 2020-01-27 at 13 49 01

After:

Screenshot 2020-01-27 at 13 45 24

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

**Comment by jasmussen** [comment]
> With this PR, the DOM tree of the columns block is exactly the same as the DOM tree on the front end. This removes the need add any styling for the editor. There's still a few things that ideally should be adjusted (we shouldn't be adding a margin to blocks by default because some blocks need to reset this). But you get the idea and the direction.

Incredible. Let's let that sink in for a moment. 

I didn't think this was possible, but it appears you may have done it. Obviously this is a big change, so it needs many eyes and reviews, and there are likely a ton of edgecases and things that will need touching up a a result of such a big refactor. But before diving into that, it seems worth listing why this matters, what the benefits are:

- The markup becomes the same. This makes the markup predictable for developers developing for the editor and the frontend.
- With the same markup front and editor, the CSS necessary for the editor style can be vastly simplified, potentially paving the way for a future where we literally load the theme CSS stylesheet into the editor, instead of loading a separate one. 
- With the same markup, front and editor, both the structure and the CSS styles for all existing core blocks can be touched up to reduce a ton of specificity, further benefitting theme developers. 

In other words, there are benefits to code quality, to developers, to themers, and to end users who will be more likely to use themes that look like the front-end, increasing usability. That's all to say that this PR is worth reviewing and helping to completion!

**Comment by paales** [comment]
This is really interesting @ellatrix! I was struggling to get webcomponents working and this might be a big step in the right direction! Maybe you can take a look at the issue (https://github.com/WordPress/gutenberg/issues/18778 ) for a second and be able to confirm/deny if your PR would solve the issue?

File: packages/block-editor/src/components/block-list/index.js [hunk 1]
```diff
@@ -90,6 +91,7 @@ function BlockList( {
 				className
 			) }
 			{ ...__experimentalContainerProps }
+			style={ style }
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/index.js]: Do you think there's a way we could remove this div or maybe pass a `Block.*` to InnerBlocks to render it as the wrapper of the current block?

Comment (ellatrix) [reply packages/block-editor/src/components/block-list/index.js]: I'm not sure I'm following... That's already the case? `Container` is variable and set by the block through `InnerBlocks`.

Comment (ellatrix) [reply packages/block-editor/src/components/block-list/index.js]: And this div element is needed and cannot be removed. Or do you mean the style prop? I do agree that all props should be passed down and not just the style prop.

Comment (youknowriad) [reply packages/block-editor/src/components/block-list/index.js]: I mean I'm a container block and I want to be a "section" and I don't want this extra div added by InnerBlocks so I was wondering if I could do:

```
<InnerBlocks tagName={ Block.section } />
```

Maybe it's already possible and there's something I'm missing

Comment (ellatrix) [reply packages/block-editor/src/components/block-list/index.js]: Like this? https://github.com/WordPress/gutenberg/pull/19910/files#diff-3878fabdd6add9b09483511e7ff0d9b3R70

Comment (youknowriad) [reply packages/block-editor/src/components/block-list/index.js]: haha I guess I missed it for some reason :)

File: packages/block-editor/src/components/inner-blocks/index.js [hunk 2]
```diff
@@ -149,26 +149,53 @@ class InnerBlocks extends Component {
...
+					'blockListSettings',
+					'parentLock',
+					'isLastBlockChangePersistent',
+					'replaceInnerBlocks',
+					'__unstableMarkNextChangeAsNotPersistent',
+					'updateNestedSettings',
+					'__experimentalMoverDirection',
+					'allowedBlocks',
+					'template',
+				] ) }
```
Comment (ellatrix) [inline packages/block-editor/src/components/inner-blocks/index.js]: This component could benefit from a hooks rewrite, which will remove the need to omit props used by this component. I'll consider it out of scope for this PR though.

File: packages/block-editor/src/components/block-list/index.js [hunk 3]
```diff
@@ -37,6 +37,9 @@ function BlockList( {
 	isDraggable,
 	renderAppender,
 	__experimentalUIParts = {},
+	tagName = 'div',
+	forwardedRef,
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/index.js]: I'm not sure I understand why we can't just use `React.forwardRef` and use just "ref" on the caller?

File: packages/block-editor/src/components/inner-blocks/index.js (hunk 2 above)
Comment (youknowriad) [reply packages/block-editor/src/components/inner-blocks/index.js]: So just to clarify here, the extra `...props` are needed for the "style" in this PR right?

File: packages/block-editor/src/components/inner-blocks/index.js [hunk 4]
```diff
@@ -149,26 +149,53 @@ class InnerBlocks extends Component {
...
+			/>
+		);
+		if ( props.tagName ) {
+			return blockList;
```
Comment (youknowriad) [inline packages/block-editor/src/components/inner-blocks/index.js]: What about `has-overlay`, `is-capturing-toolbar` and `block-editor-inner-blocks` classNames, don't we need them when we provide a tagName?

Also why the extra "div" when we don't provide a tagName? Do you think it can just be a fallback value for "tagName"?

File: packages/block-library/src/column/edit.js [hunk 5]
```diff
@@ -61,11 +64,14 @@ function ColumnEdit( {
 				templateLock={ false }
 				renderAppender={
 					hasChildBlocks
-						? undefined
+						? false
 						: () => <InnerBlocks.ButtonBlockAppender />
 				}
+				tagName={ Block.div }
+				className={ classes }
+				style={ hasWidth ? { flexBasis: width + '%' } : undefined }
```
Comment (youknowriad) [inline packages/block-library/src/column/edit.js]: I'm wonder if instead of having these extra props here we ccould have a way to include these in the tagName

```
function Column = ( props ) => <Block.div {...props } className={ classnames( props.className, classes ) } style={ someStyle } /> 
```

And use `Column` as a tag name. 

This could be a way to limit the public API. Thoughts? too hard to understand?

File: packages/block-editor/src/components/block-list/index.js (hunk 3 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/index.js]: Yes, sorry, we should do that :)

File: packages/block-library/src/columns/edit.js [hunk 6]
```diff
@@ -103,12 +104,13 @@ function ColumnsEditContainer( {
...
-					<div className={ classes } ref={ ref }>
-						<InnerBlocks
-							allowedBlocks={ ALLOWED_BLOCKS }
-							__experimentalMoverDirection="horizontal"
-						/>
-					</div>
+					<InnerBlocks
+						allowedBlocks={ ALLOWED_BLOCKS }
+						__experimentalMoverDirection="horizontal"
+						tagName={ Block.div }
```
Comment (youknowriad) [inline packages/block-library/src/columns/edit.js]: I believe the Block component automatically adds the "className" prop of the block right? Should we remove it from the "classes" variable then? (same for the other block)

File: packages/block-editor/src/components/inner-blocks/index.js (hunk 2 above)
Comment (ellatrix) [reply packages/block-editor/src/components/inner-blocks/index.js]: That's correct. Any remaining props on `InnerBlocks` should be passed down to the block wrapper.

File: packages/block-editor/src/components/inner-blocks/index.js (hunk 4 above)
Comment (ellatrix) [reply packages/block-editor/src/components/inner-blocks/index.js]: Yes, either we should merge #19849 first, or just incorporate it into this PR.

> Also why the extra "div" when we don't provide a tagName?

For back compat... I don't want to invalidate styling for existing blocks using `InnerBlocks` without a tag name.

File: packages/block-library/src/column/edit.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-library/src/column/edit.js]: That could work. Downside is that this component must be created on every render, but that's probably fine if it's simple.

File: packages/block-library/src/columns/edit.js (hunk 6 above)
Comment (ellatrix) [reply packages/block-library/src/columns/edit.js]: Not sure what you mean. Remove the next line? It's passing some custom classes too?

Comment (youknowriad) [reply packages/block-library/src/columns/edit.js]: Question first: Block.div is aware of the block type... so theorically we can absorb the need to apply the "className" prop there so block authors don't forget about it anymore. Is this already the case?

Comment (ellatrix) [reply packages/block-library/src/columns/edit.js]: It doesn't look like we currently do. Only these classes are "enforced":

https://github.com/WordPress/gutenberg/blob/d6b5e8e5b256884fb0378424a6fd87f84565fa1c/packages/block-editor/src/components/block-list/block.js#L98-L118

I'm fine with enforcing the block type class, as long as we're doing the same on the front-end. Are we enforcing it on the front-end?

Comment (youknowriad) [reply packages/block-library/src/columns/edit.js]: yes we are :) Seems like a good separate PR.

File: packages/block-editor/src/components/block-list/index.js [hunk 7]
```diff
@@ -136,7 +129,16 @@ function BlockList( {
...
 }
 
+const ForwardedBlockList = forwardRef( BlockList );
+
 // This component needs to always be synchronous
 // as it's the one changing the async mode
 // depending on the block selection.
-export default forceSyncUpdates( BlockList );
+export default forwardRef( ( props, ref ) => {
+	const fallbackRef = useRef();
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/index.js]: This seems like something that could be built-in `forwardRef` with an option or something.

Comment (ellatrix) [reply packages/block-editor/src/components/block-list/index.js]: I'd love that. We're doing it in other components. Are we ok with enhancing a React API in `@wordpress/element`?

Comment (youknowriad) [reply packages/block-editor/src/components/block-list/index.js]: I don't know, maybe we could ask the React team too if this is a pattern they've noticed.

Comment (ellatrix) [reply packages/block-editor/src/components/block-list/index.js]: Sounds good! Let's discuss separately.

File: packages/block-library/src/column/edit.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-library/src/column/edit.js]: Thinking about it more, I don't like that it's now the responsibility of the block to pass `props` and to make sure that these props are properly merged. If we need to isolate these props, perhaps it's better to nest them in an object. I don't have anything against that, but I don't have anything against not nesting them either. What do you think?

**Comment by ellatrix** [comment]
Fixed the e2e tests and it's looking good to me now.

File: packages/block-library/src/column/edit.js (hunk 5 above)
Comment (youknowriad) [reply packages/block-library/src/column/edit.js]: Would you mind sharing a code example to clarify?

Comment (ellatrix) [reply packages/block-library/src/column/edit.js]: For nesting props?

```
<InnerBlocks
	templateLock={ false }
	renderAppender={
		hasChildBlocks
			? false
			: () => <InnerBlocks.ButtonBlockAppender />
	}
	tagName={ Block.div }
	passedProps={ {
		className: classes,
		style: hasWidth ? { flexBasis: width + '%' } : undefined,
	} }
/>
```

Comment (ellatrix) [reply packages/block-library/src/column/edit.js]: Btw, we have a similar situation for `RichText`. Probably would be good to stay consistent.

Comment (youknowriad) [reply packages/block-library/src/column/edit.js]: Right, how do we do with RichText, just spread props?

Comment (ellatrix) [reply packages/block-library/src/column/edit.js]: Yeah, we spread just like this PR atm.

Comment (ellatrix) [reply packages/block-library/src/column/edit.js]: @youknowriad What do you think? Spread, nested object, or pass component?

Comment (youknowriad) [reply packages/block-library/src/column/edit.js]: Honestly, I don't know yet :P Maybe one of "pass component" or "experimental nested object" to avoid having to make the decision now and explore more blocks first.

Comment (ellatrix) [reply packages/block-library/src/column/edit.js]: Agreed.

Comment (ellatrix) [reply packages/block-library/src/column/edit.js]: I think I'll go for experimentally nesting the props in an object. A new component becomes quite complex as you'd also have to forward the ref. :/

**Comment by ellatrix** [comment]
This is working very well. Let's merge and iterate.

File: packages/block-editor/src/utils/dom.js [hunk 8]
```diff
@@ -76,7 +76,10 @@ export function isInsideRootBlock( blockElement, element ) {
  * @return {boolean} Whether element contains inner blocks.
  */
 export function hasInnerBlocksContext( element ) {
-	return !! element.querySelector( '.block-editor-block-list__layout' );
+	return (
+		element.classList.contains( 'block-editor-block-list__layout' ) ||
+		!! element.querySelector( '.block-editor-block-list__layout' )
+	);
```
Comment (aduth) [inline packages/block-editor/src/utils/dom.js]: There are tests for this function. Probably should have been updated.

File: packages/block-library/src/column/edit.js [hunk 9]
```diff
@@ -61,11 +64,16 @@ function ColumnEdit( {
 				templateLock={ false }
 				renderAppender={
 					hasChildBlocks
-						? undefined
+						? false
 						: () => <InnerBlocks.ButtonBlockAppender />
 				}
+				__experimentalTagName={ Block.div }
```
Comment (aduth) [inline packages/block-library/src/column/edit.js]: It wasn't introduced here, but it's not clear to me why we limit to a predefined set of tag names, vs. some sort of factor function (`createBlockComponent( tagName )`) or just rendering some element like `<BlockComponent tagName="div" />` here. What if I wanted a `section` block?

File: packages/block-library/src/column/index.js [hunk 10]
```diff
@@ -24,17 +24,7 @@ export const settings = {
...
-		const { width } = attributes;
-		if ( Number.isFinite( width ) ) {
-			return {
-				style: {
-					flexBasis: width + '%',
-				},
-				'data-has-explicit-width': true,
-			};
-		}
+		lightBlockWrapper: true,
```
Comment (aduth) [inline packages/block-library/src/column/index.js]: Is this experimental? Where do we document it?

File: packages/block-library/src/columns/style.scss [hunk 11]
```diff
@@ -56,7 +56,7 @@
 
 		// Columns with an explicitly-assigned width should maintain their
 		// `flex-basis` width and not grow.
-		&[style] {
+		&[style*="flex-basis"] {
```
Comment (aduth) [inline packages/block-library/src/columns/style.scss]: Are we now applying other styles? This was already pretty fragile in the first place, but was a compromise since existing content wouldn't have any indicators of whether a predefined width was assigned. I think if we expand this, we might want to consider deprecations to include some more durable indicator (like a `data-` attribute, or class name).

File: packages/block-library/src/column/edit.js (hunk 9 above)
Comment (ellatrix) [reply packages/block-library/src/column/edit.js]: It's still experimental, but the intention is to add all tags.

File: packages/block-library/src/column/index.js (hunk 10 above)
Comment (ellatrix) [reply packages/block-library/src/column/index.js]: Ah, it should have been prefixed as well with `__experimental`.

File: packages/block-library/src/columns/style.scss (hunk 11 above)
Comment (ellatrix) [reply packages/block-library/src/columns/style.scss]: Yes, there are other inline styles for block animations. Yes, would probably be best to have a separate attribute.

**Comment by jasmussen** [comment]
This PR had a side-effect that may have been intended, not sure.

Before this PR, the last block in a nesting container would show a big visible plus button, like this:

Screenshot 2020-03-10 at 14 41 38

The heuristic was a bit overthought, perhaps — it would show up at the end of a blocklist if the last block was not a paragraph. 

After this PR, that plus is gone. Technically, I find this to be a vastly better experience because that plus should disappear until a block is selected.

However it would be good to find out if there were any ill effects from the removal of these. And it'd also be good to know exactly _why_ it was removed — i.e. was it unintentional and is there dead code somewhere? 

CC: @ellatrix

**Comment by ellatrix** [comment]
@jasmussen It looks like this was an accidental removal:

https://github.com/WordPress/gutenberg/pull/19910/files#diff-3878fabdd6add9b09483511e7ff0d9b3L64

Explicitly setting this to false won't render any appender, and undefined seems to only render it if the last block is not a paragraph. It's been set to false because, even if there was no appender, some invisible DOM would be left that messes up the :last-child rules. I didn't know something would be rendered if the last block was not a paragraph. Should it be restored?

**Comment by jasmussen** [comment]
Excellent detective work.

We should not restore it until we have a good plan for what to do with it. Knowing how to add it back thanks to you help, I might try my hand at a PR tomorrow.

**Comment by ZebulanStanphill** [comment]
@cr0ybot #20650, #20656, and #20689 may be relevant to you.
